Source/WebKit/GPUProcess/media/cocoa/RemoteMediaPlayerManagerProxyCocoa.mm

```diff
@@ -0,0 +1,108 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "RemoteMediaPlayerManagerProxy.h"
+
+#if ENABLE(GPU_PROCESS) && ENABLE(VIDEO)
+
+#include "VideoReceiverEndpointMessage.h"
+#include <WebCore/MediaPlayerPrivate.h>
+#include <WebCore/VideoTargetFactory.h>
+
+namespace WebKit {
+
+#if ENABLE(LINEAR_MEDIA_PLAYER)
+PlatformVideoTarget RemoteMediaPlayerManagerProxy::videoTargetForMediaElementIdentifier(WebCore::HTMLMediaElementIdentifier mediaElementIdentifier)
+{
+    auto cachedEntry = m_videoReceiverEndpointCache.find(mediaElementIdentifier);
+    if (cachedEntry == m_videoReceiverEndpointCache.end())
+        return nullptr;
+
+    return cachedEntry->value.videoTarget;
+}
+
+void RemoteMediaPlayerManagerProxy::handleVideoReceiverEndpointMessage(const VideoReceiverEndpointMessage& endpointMessage)
+{
+    auto cachedEntry = m_videoReceiverEndpointCache.find(endpointMessage.mediaElementIdentifier());
+    if (cachedEntry == m_videoReceiverEndpointCache.end()) {
+        // If no entry for the specified mediaElementIdentifier exists, add a new entry to
+        // the cache, and set the new endpoint on the specified MediaPlayer.
+        auto videoTarget = WebCore::VideoTargetFactory::createTargetFromEndpoint(endpointMessage.endpoint());
+        m_videoReceiverEndpointCache.set(endpointMessage.mediaElementIdentifier(), VideoRecevierEndpointCacheEntry { endpointMessage.playerIdentifier(), endpointMessage.endpoint(), videoTarget });
+
+        if (RefPtr mediaPlayer = this->mediaPlayer(endpointMessage.playerIdentifier()))
+            mediaPlayer->setVideoTarget(videoTarget);
+
+        return;
+    }
+
+    // An cached entry implies a MediaPlayer has already been given this endpoint.
+    auto cachedPlayerIdentifier = cachedEntry->value.playerIdentifier;
+    auto cachedEndpoint = cachedEntry->value.endpoint;
+
+    // If nothing has actually changed, bail.
+    if (cachedPlayerIdentifier == endpointMessage.playerIdentifier()
+        && cachedEndpoint.get() == endpointMessage.endpoint().get())
```

Comment:
This check seems harmless, but I don't think it's possible for `cachedEndpoint` to ever equal `endpointMessage.endpoint()` since there will be a new instance of an `xpc_object_t` each time `endpointMessage` is deserialized.